### PR TITLE
Combine VB comparers into one, and combine VB and C# comparers together

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/51169

Apologies this is going to be a pain to review, as its pretty much all just code moves.

### CODE REVIEW

**Comment by davidwengier** [comment]
One test failed. That's very surprising. Figured I would have made a mistake and most of this would be on the floor, or I would have done it right and it would all be green. Only one is odd. Will look into it on Monday.

File: src/Features/CSharp/Portable/EditAndContinue/AbstractSyntaxComparer.cs [hunk 1]
```diff
@@ -6,34 +6,22 @@
 using System.Collections.Immutable;
 using System.Diagnostics;
 using Microsoft.CodeAnalysis.Differencing;
-using Microsoft.CodeAnalysis.Text;
 namespace Microsoft.CodeAnalysis.CSharp.EditAndContinue
 {
-    internal abstract class AbstractSyntaxComparer : TreeComparer<SyntaxNode>
+    internal abstract class AbstractSyntaxComparer : Microsoft.CodeAnalysis.Differencing.AbstractSyntaxComparer
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/AbstractSyntaxComparer.cs:11]: >AbstractSyntaxComparer [](start = 28, length = 22)

Do we still have multiple subclasses of the C# comparer? If not we should merge the SyntaxComparer with this AbstractSyntaxComparer

File: src/Features/VisualBasic/Portable/EditAndContinue/SyntaxComparer.vb [hunk 2]
```diff
@@ -2,230 +2,1300 @@
...
             End If
 
-            Dim distance As Double = ComputeDistance(leftNode, rightNode)
-            Return If(distance = ExactMatchDist, EpsilonDist, distance)
+            Return ClassifyTopSytnax(kind, isLeaf, ignoreVariableDeclarations)
         End Function
 
-        Friend Overloads Shared Function ComputeDistance(oldNodeOrToken As SyntaxNodeOrToken, newNodeOrToken As SyntaxNodeOrToken) As Double
-            Debug.Assert(newNodeOrToken.IsToken = oldNodeOrToken.IsToken)
+        Friend Shared Function ClassifyStatementSytnax(kind As SyntaxKind, nodeOpt As SyntaxNode, ByRef isLeaf As Boolean) As Label
```
Comment (tmat) [inline src/Features/VisualBasic/Portable/EditAndContinue/SyntaxComparer.vb:327]: >ClassifyStatementSytnax [](start = 31, length = 23)

typo: Syntax

File: src/Features/VisualBasic/Portable/EditAndContinue/AbstractSyntaxComparer.vb [hunk 3]
```diff
@@ -0,0 +1,212 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
+' The .NET Foundation licenses this file to you under the MIT license.
+' See the LICENSE file in the project root for more information.
+
+Imports System.Collections.Immutable
+Imports Microsoft.CodeAnalysis.Differencing
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.EditAndContinue
+
+    Friend MustInherit Class AbstractSyntaxComparer
```
Comment (tmat) [inline src/Features/VisualBasic/Portable/EditAndContinue/AbstractSyntaxComparer.vb:10]: >AbstractSyntaxComparer [](start = 29, length = 22)

Same here - are there multiple subclasses?

File: src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs [hunk 4]
```diff
@@ -47,178 +25,17 @@ public SyntaxComparer(bool compareStatementSyntax)
...
         public SyntaxComparer(
-            SyntaxNode oldRoot,
-            SyntaxNode newRoot,
-            IEnumerable<SyntaxNode> oldRootChildren,
-            IEnumerable<SyntaxNode> newRootChildren,
+            SyntaxNode? oldRoot,
+            SyntaxNode? newRoot,
+            IEnumerable<SyntaxNode>? oldRootChildren,
+            IEnumerable<SyntaxNode>? newRootChildren,
             bool compareStatementSyntax = false)
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs:32]: >false [](start = 42, length = 5)

i'd make this a equire parameter.

**Review by tmat — CHANGES_REQUESTED** [review]
:clock1:

File: src/Features/CSharp/Portable/EditAndContinue/AbstractSyntaxComparer.cs (hunk 1 above)
Comment (davidwengier) [reply src/Features/CSharp/Portable/EditAndContinue/AbstractSyntaxComparer.cs:11]: No, I meant to mention this at the top but forgot. Your suggestion is one of the potential solutions though, and the one I was leaning towards, so I'll go with it :)

**Comment by davidwengier** [comment]
@tmat this should be good to go?

**Review by tmat — APPROVED** [review]
:shipit:

**Comment by tmat** [comment]
@davidwengier Yup, let's merge.

**Pull request merged**